### ISSUE TITLE
Running "meson configure" in source directory raises an exception (Episode IV: The AssertionError Strikes Back)

### ISSUE THREAD

**Issue by Akaricchi** [mesonbuild/meson#6748]
**Describe the bug**
```
Traceback (most recent call last):
  File "/data/git/meson/mesonbuild/mesonmain.py", line 131, in run
    return options.run_func(options)
  File "/data/git/meson/mesonbuild/mconf.py", line 257, in run
    c = Conf(builddir)
  File "/data/git/meson/mesonbuild/mconf.py", line 64, in __init__
    intr.analyze()
  File "/data/git/meson/mesonbuild/ast/introspection.py", line 303, in analyze
    self.parse_project()
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 414, in parse_project
    self.evaluate_codeblock(self.ast, end=1)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 454, in evaluate_codeblock
    raise e
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 447, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 459, in evaluate_statement
    return self.function_call(cur)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 818, in function_call
    return func(node, func_args, self.kwargs_string_keys(kwargs))
  File "/data/git/meson/mesonbuild/ast/introspection.py", line 120, in func_project
    self.do_subproject(i)
  File "/data/git/meson/mesonbuild/ast/introspection.py", line 133, in do_subproject
    subi.analyze()
  File "/data/git/meson/mesonbuild/ast/introspection.py", line 304, in analyze
    self.run()
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 429, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 454, in evaluate_codeblock
    raise e
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 447, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 461, in evaluate_statement
    self.assignment(cur)
  File "/data/git/meson/mesonbuild/ast/interpreter.py", line 260, in assignment
    self.assign_vals[node.var_name] = self.evaluate_statement(node.value) # Evaluate the value just in case
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 477, in evaluate_statement
    return self.evaluate_dictstatement(cur)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 268, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 523, in evaluate_dictstatement
    assert isinstance(key, mparser.BaseNode)  # All keys must be nodes due to resolve_key_nodes=False
AssertionError
```

**To Reproduce**
I don't have a minimal failure case, but this is known to affect two projects so far:
* [Taisei Project](https://github.com/taisei-project/taisei)
* [My custom SDL2 meson port](https://github.com/taisei-project/SDL-mirror/tree/meson-2.0.10-emscripten-fixes)

Similar bugs [have](https://github.com/mesonbuild/meson/issues/4960) [been](https://github.com/mesonbuild/meson/issues/5256) [found](https://github.com/mesonbuild/meson/issues/5909) by Taisei in the past; I wouldn't mind if you'd just include the whole project in your integration test suite or something.

**Expected behavior**
A list of project's options printed out.

**system parameters**
* openSUSE Tumbleweed
* Python 3.7.3
* Meson 0.53.999 (74769617907f571d2099001d3a7443e23b4f6cda)

**Comment by Akaricchi** [comment]
cc @mensinda

**Comment by xclaesse** [comment]
Could you point to which line in your meson.build that issue comes from, please? That must be when parsing a dict `{  }`

**Comment by xclaesse** [comment]
Also which version of Meson? There have been changes how meson parse dict keys in 0.53.0.

**Comment by Akaricchi** [comment]
> Could you point to which line in your meson.build that issue comes from, please? That must be when parsing a dict { }

My `meson.build`s don't have any dicts. How do I print out the file:line?

> Also which version of Meson? There have been changes how meson parse dict keys in 0.53.0.

See the main post
> Meson 0.53.999 (7476961)

**Comment by xclaesse** [comment]
Please post the full logs so we can see roughly where it was when that happened. Pretty sure it has a dict in a subproject, judging by the backtrace.

**Comment by Akaricchi** [comment]
Sorry, I forgot there are actually dicts in the SDL2 subproject. The backtrace for just that one looks like this:

```
Traceback (most recent call last):
  File "/data/git/meson/mesonbuild/mesonmain.py", line 131, in run
    return options.run_func(options)
  File "/data/git/meson/mesonbuild/mconf.py", line 257, in run
    c = Conf(builddir)
  File "/data/git/meson/mesonbuild/mconf.py", line 64, in __init__
    intr.analyze()
  File "/data/git/meson/mesonbuild/ast/introspection.py", line 304, in analyze
    self.run()
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 429, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 454, in evaluate_codeblock
    raise e
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 447, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 461, in evaluate_statement
    self.assignment(cur)
  File "/data/git/meson/mesonbuild/ast/interpreter.py", line 260, in assignment
    self.assign_vals[node.var_name] = self.evaluate_statement(node.value) # Evaluate the value just in case
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 477, in evaluate_statement
    return self.evaluate_dictstatement(cur)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 268, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/data/git/meson/mesonbuild/interpreterbase.py", line 523, in evaluate_dictstatement
    assert isinstance(key, mparser.BaseNode)  # All keys must be nodes due to resolve_key_nodes=False
AssertionError
```

The one I initially posted is from Taisei, which includes SDL2 as a subproject. I've also tried it on a fresh clone of Taisei (which doesn't have dicts), **without** checking out any subprojects, and got a surprising result: **`meson configure` just silently exits with return code `1`**, no traceback or anything. This may be a different bug.

I'm not sure where the logs would be. There is no build directory, I'm just running it on the source dir. So there's no `meson-logs/`, either.

**Comment by mensinda** [comment]
The return code 1 issue should be fixed with #6751.
